src/mlpack/tests/rectangle_tree_test.cpp

```diff
@@ -854,6 +864,244 @@ BOOST_AUTO_TEST_CASE(DiscreteHilbertValueTest)
                                                                   point4), -1);
 }
 
+template<typename TreeType>
+void CheckOverlap(const TreeType& tree)
+{
+  bool success = true;
+
+  // Check if two nodes overlap each other.
+  for (size_t i = 0; i < tree.NumChildren(); i++)
+  {
+    success = true;
+
+    for (size_t j = 0; j < tree.NumChildren(); j++)
+    {
+      if (j == i)
+        continue;
+      success = false;
+      // Two nodes overlap each other if and only if there are no dimension
+      // in which they do not overlap each other.
+      for (size_t k = 0; k < tree.Bound().Dim(); k++)
+      {
+        if ((tree.Child(i).Bound()[k].Lo() >=
+            tree.Child(j).Bound()[k].Hi()) ||
+            (tree.Child(j).Bound()[k].Lo() >=
+            tree.Child(i).Bound()[k].Hi()))
+        {
+          success = true;
+          break;
```

Comment:
I guess `HRectBound::Intersect` or `HRectBound::Overlap` probably is a better name, how do you think?
Moreover, `r_start_tree_split_impl.hpp`, `x_tree_split_imp.hpp` and `minimal_coverage_sweep_impl.hpp` contain a piece of almost equal code. I mean the code that computes the total coverage of two sets of nodes (points) and the area of intersection.
The code is slightly different in `MinimalCoverageSweep` (it takes into account that bounds can be empty). Maybe we should add this functionality to `HRectBound`. (something like this: `HRectBound::MeasureSplitCoverageOverlap(TreeType* node, SortStruct& sorted, size_t splitIndex, ElemType& area1, ElemType& area2, ElemType& overlappedArea)`)